packages/oauth/src/providers/keycloak.ts

```diff
@@ -0,0 +1,232 @@
+import {
+	OAuth2ProviderAuthWithPKCE,
+	createOAuth2AuthorizationUrlWithPKCE,
+	validateOAuth2AuthorizationCode
+} from "../core/oauth2.js";
+import { ProviderUserAuth } from "../core/provider.js";
+import { decodeIdToken } from "../index.js";
+import { encodeBase64 } from "../utils/encode.js";
+import { handleRequest, authorizationHeader } from "../utils/request.js";
+
+import type { Auth } from "lucia";
+
+type Config = {
+  domain: string;
+  realm: string;
+  clientId: string;
+  clientSecret: string;
+  scope?: string[];
+  redirectUri?: string;
+};
+
+const PROVIDER_ID = "keycloak";
+
+export const keycloak = <_Auth extends Auth = Auth>(
+	auth: _Auth,
+	config: Config
+): KeycloakAuth<_Auth> => {
+	return new KeycloakAuth(auth, config);
+};
+
+export class KeycloakAuth<_Auth extends Auth = Auth> extends OAuth2ProviderAuthWithPKCE<
+	KeycloakUserAuth<_Auth>
+> {
+	private config: Config;
+
+	constructor(auth: _Auth, config: Config) {
+		super(auth);
+
+		this.config = config;
+	}
+
+	public getAuthorizationUrl = async (): Promise<
+		readonly [url: URL, codeVerifier: string, state: string]
+	> => {
+    const scopeConfig = this.config.scope ?? [];
+		return await createOAuth2AuthorizationUrlWithPKCE(
+			`https://${ this.config.domain }/realms/${ this.config.realm }/protocol/openid-connect/auth`,
+			{
+        clientId: this.config.clientId,
+        scope: ["profile", "openid", ...scopeConfig],
+        redirectUri: this.config.redirectUri,
+        codeChallengeMethod: "S256"
+			}
+		);
+	};
+
+	public validateCallback = async (
+		code: string,
+    code_verifier: string
+	): Promise<KeycloakUserAuth<_Auth>> => {
+		const keycloakTokens = await this.validateAuthorizationCode(code, code_verifier);
+		const keycloakUser = await getKeycloakUser(this.config.domain, this.config.realm, keycloakTokens.accessToken);
+		const keycloakRoles = getKeycloakRoles(keycloakTokens.accessToken)
+		return new KeycloakUserAuth(this.auth, keycloakUser, keycloakTokens, keycloakRoles);
+	};
+
+	private validateAuthorizationCode = async (
+		code: string,
+    codeVerifier: string
+	): Promise<KeycloakTokens> => {
+		const rawTokens = await validateOAuth2AuthorizationCode<AccessTokenResponseBody>(
+			code,
+			`https://${ this.config.domain }/realms/${ this.config.realm }/protocol/openid-connect/token`,
+			{
+				clientId: this.config.clientId,
+				redirectUri: this.config.redirectUri,
+				codeVerifier,
+				clientPassword: {
+					authenticateWith: "http_basic_auth",
+					clientSecret: this.config.clientSecret
+				}
+			}
+		);
+
+		return this.claimTokens(rawTokens)
+	};
+
+	private claimTokens = (tokens: AccessTokenResponseBody): KeycloakTokens => {
+		const tokenDecoded = decodeIdToken<Claims>(tokens.access_token)
+
+		if ("refresh_token" in tokens) {
+			return {
+				accessToken: tokens.access_token,
+				accessTokenExpiresIn: tokens.expires_in,
+				authTime: tokenDecoded.auth_time,
```

Comment:
I think it's better if you just return the ID token here? Not sure why you need both `accessTokenExpiresIn` and `expiresIn`?

Reply:
There is accesTokenExpiresIn with duration till expration and there ist expires**At** with the exact time of expiration.
We claim everywhere the expiration times, why shouldnt we complete these informations if we can and just return the ID token?